experimental/pattern-matching.md

```diff
@@ -0,0 +1,119 @@
+# [Pattern Matching](proposal-pattern-matching)
+
+See also [AST examples](ast-examples) for this proposal.
+
+## MatchExpression
+
+```js
+interface MatchExpression <: Expression {
+    type: "MatchExpression";
+    discriminant: Expression;
+    id: Pattern | null;
+    clause: [ MatchClause ]
+}
+```
+
+## WhenClause
+```js
+interface MatchClause <: Node {
+    type: "MatchClause";
+    test: MatchPattern | null;
+    guard: Expression | null;
+    id: Pattern | null;
+    consequent: BlockStatement
+}
+```
+
+If `test` is `null`, `id` must be `null`.
+
+## MatchPattern
+```js
+interface MatchPattern <: Node {}
+```
+
+## ArrayMatchPattern
+```js
+interface ArrayMatchPattern <: MatchPattern {
+    type: "ArrayMatchPattern";
+    elements: [ MatchPattern ];
+}
+```
+
+## ObjectMatchPattern
+```js
+interface AssignmentMatchProperty <: Property {
+    type: "Property"; // inherited
+    value: MatchPattern;
+    kind: "init";
+    method: false;
+}
+
+interface ObjectMatchPattern <: MatchPattern {
+    type: "ObjectMatchPattern";
+    properties: [ AssignmentMatchProperty ];
+}
+```
+
+## RestMatchElement
+```js
+interface RestMatchElement <: MatchPattern {
+    type: "RestMatchElement";
+    argument: MatchPattern;
+}
+```
+
+## BinaryMatchPattern
+```js
+interface BinaryMatchPattern <: MatchPattern {
+    type: "BinaryPattern";
+    operator: "and" | "or" | "with";
+    left: MatchPattern;
+    right: MatchPattern;
+}
+```
+
+## AsMatchPattern
+```js
+interface AsMatchPattern <: MatchPattern {
+    type: "AsMatchPattern";
+    test: MatchPattern;
+    id: Pattern;
+}
+```
+
+## ExpressionMatchPattern
+```js
+interface ExpressionMatchPattern <: MatchPattern {
+    type: "ExpressionMatchPattern";
+    expression: Expression;
+}
+```
+
+## NullMatchPattern
+```js
+interface NullMatchPattern <: MatchPattern {
+    type: "NullMatchPattern";
+}
+```
+
+# Literal
+
+```js
+extend interface Literal <: Expression, MatchPattern {}
+```
+
+# Identifier
+```js
+extend interface Identifier <: Expression, Pattern, MatchPattern {}
+```
+
+### UnaryExpression
+
+```js
+extend interface UnaryExpression <: Expression, MatchPattern {}
+```
+
+If a UnaryExpression is a MatchPattern, its `operator` must be `"-"`, its `prefix` must be `true`, and its `argument` must be either a Literal with numeric value, or an identifier named `"Infinity"`.
```

Comment:
as a pattern, this also allows `undefined` (even if the identifier resolves to something that's not the `undefined` value), but also, all number and bigint literals (with +, -, or no sign at all), `null`, booleans, etc.

Also, template literals haven't been thought out here - i filed https://github.com/tc39/proposal-pattern-matching/issues/205 to track

Reply:
Yeah, `undefined` is covered in the `Identifier` case

```
extend interface Identifier <: Expression, Pattern, MatchPattern {}
```

The `UnaryExpression` case is addressed to things like `-0`, `-1`, `-1n`, since they are an UnaryExpression from the AST perspective.

The template literals in ESTree has a specific type `TemplateLiteral`, in this PR the `TemplateLiteral` is not mentioned so it is not a `MatchPattern`. But yeah it is a good question whether it should be allowed. I will update this PR if necessary.